### PR TITLE
Rework how DeterministicRunnerImpl and SyncWorkflowContext handle creation of workflow root and method threads

### PR DESCRIPTION
**Pull request by Spikhalskiy**
Refactored workflow root thread into a separate code path making it explicitly different from workflow-method thread and child workflows.
Added WorkflowInboundCallsInterceptor#newWorkflowMethodThread intercepting creation of the main workflow-method thread.
Added WorkflowInboundCallsInterceptor#newCallbackThread intercepting creation of a callback thread.
WorkflowOutboundCallsInterceptor#newThread is renamed into WorkflowOutboundCallsInterceptor#newChildThread and now it intercepts only new workflow child threads and doesn't intercept creation of the main workflow-method thread.
Fixed situation that the same "workflow-method" name was used for workflow initializer thread (now called "workflow-root") and for the thread that is used to run a workflow method (continued to be "workflow-method").

This PR is a first step to address issue #537

### CODE REVIEW

File: temporal-sdk/src/main/java/io/temporal/common/interceptors/WorkflowInboundCallsInterceptor.java [hunk 1]
```diff
@@ -128,4 +130,22 @@ public Object getResult() {
 
   /** Called when a workflow is queried. */
   QueryOutput handleQuery(QueryInput input);
+
+  /**
+   * Intercepts creation of the workflow main method thread
+   *
+   * @param runnable thread function to run
+   * @param name name of the thread, optional
+   * @return created WorkflowThread
```
Comment (mfateev) [inline temporal-sdk/src/main/java/io/temporal/common/interceptors/WorkflowInboundCallsInterceptor.java:139]: Change comment to smth like:

@return created workflow thread. Treat the result as an opaque object that shouldn't be manipulated in any way by the interceptor code.

File: temporal-sdk/src/main/java/io/temporal/common/interceptors/WorkflowOutboundCallsInterceptor.java [hunk 2]
```diff
@@ -483,7 +478,15 @@ public DynamicSignalHandler getHandler() {
 
   void upsertSearchAttributes(Map<String, Object> searchAttributes);
 
-  Object newThread(Runnable runnable, boolean detached, String name);
+  /**
+   * Intercepts creation of the workflow child thread
```
Comment (mfateev) [inline temporal-sdk/src/main/java/io/temporal/common/interceptors/WorkflowOutboundCallsInterceptor.java:482]: I fear that can get confused with child workflows. Maybe add to the comment that it is not related to child workflows in any way.

File: temporal-sdk/src/main/java/io/temporal/internal/sync/DeterministicRunnerImpl.java [hunk 3]
```diff
@@ -360,10 +313,10 @@ public void close() {
       return;
     }
     try {
-      for (WorkflowThread c : threadsToAdd) {
+      for (WorkflowThread c : workflowThreadsToAdd) {
         threads.add(c);
       }
-      threadsToAdd.clear();
+      workflowThreadsToAdd.clear();
```
Comment (mfateev) [inline temporal-sdk/src/main/java/io/temporal/internal/sync/DeterministicRunnerImpl.java:319]: Should we do the same for the callback threads to add?

File: temporal-sdk/src/main/java/io/temporal/internal/sync/SyncWorkflowContext.java [hunk 4]
```diff
@@ -139,15 +141,29 @@ public DeterministicRunner getRunner() {
...
-      this.headInterceptor = head;
+  public WorkflowInboundCallsInterceptor getWorkflowInboundInterceptor() {
+    // This is needed for unit tests that create DeterministicRunner directly.
+    return headInboundInterceptor == null
+        ? new BaseRootWorkflowInboundCallsInterceptor(this)
+        : headInboundInterceptor;
+  }
+
+  public void initHeadOutboundCallsInterceptor(WorkflowOutboundCallsInterceptor head) {
+    if (headOutboundInterceptor == null) {
```
Comment (mfateev) [inline temporal-sdk/src/main/java/io/temporal/internal/sync/SyncWorkflowContext.java:157]: Why this check?

File: temporal-sdk/src/main/java/io/temporal/internal/sync/SyncWorkflowContext.java [hunk 5]
```diff
@@ -139,15 +141,29 @@ public DeterministicRunner getRunner() {
...
+      headOutboundInterceptor = head;
+    }
+  }
+
+  public void initHeadInboundCallsInterceptor(WorkflowInboundCallsInterceptor head) {
+    if (headInboundInterceptor == null) {
```
Comment (mfateev) [inline temporal-sdk/src/main/java/io/temporal/internal/sync/SyncWorkflowContext.java:163]: Why null check?

File: temporal-sdk/src/main/java/io/temporal/common/interceptors/WorkflowInboundCallsInterceptor.java [hunk 6]
```diff
@@ -128,4 +130,22 @@ public Object getResult() {
...
+   */
+  Object newMethodThread(Runnable runnable, @Nullable String name);
```
Comment (mfateev) [inline temporal-sdk/src/main/java/io/temporal/common/interceptors/WorkflowInboundCallsInterceptor.java:141]: Rename to newWorkflowMethodThread or newMainWorkflowThread.

File: temporal-sdk/src/main/java/io/temporal/internal/sync/WorkflowThreadImpl.java [hunk 7]
```diff
@@ -45,7 +45,7 @@
 import org.slf4j.LoggerFactory;
 import org.slf4j.MDC;
 
-class WorkflowThreadImpl implements WorkflowThread {
+public class WorkflowThreadImpl implements WorkflowThread {
```
Comment (mfateev) [inline temporal-sdk/src/main/java/io/temporal/internal/sync/WorkflowThreadImpl.java:48]: Why public?

File: temporal-sdk/src/test/java/io/temporal/internal/sync/PromiseTest.java [hunk 8]
```diff
@@ -55,11 +55,10 @@ public void testFailure() throws Throwable {
             () -> {
               CompletablePromise<Boolean> f = Workflow.newPromise();
               trace.add("root begin");
-              WorkflowInternal.newThread(
-                      false, () -> f.completeExceptionally(new IllegalArgumentException("foo")))
+              WorkflowThread.newThread(
```
Comment (mfateev) [inline temporal-sdk/src/test/java/io/temporal/internal/sync/PromiseTest.java:58]: The idea of WorkflowInternal is to proxy all the public APIs through it and keep all the implementation details private. Making WorkflowThread and its impl public goes against this idea.

File: temporal-sdk/src/main/java/io/temporal/internal/sync/DeterministicRunnerImpl.java (hunk 3 above)
Comment (Spikhalskiy) [reply temporal-sdk/src/main/java/io/temporal/internal/sync/DeterministicRunnerImpl.java:319]: It's actually what I did in this PR, I made a collection `callbackThreadsToAdd` that works almost the same way (just traversed in a reversed order when we add)

File: temporal-sdk/src/main/java/io/temporal/internal/sync/SyncWorkflowContext.java (hunk 5 above)
Comment (Spikhalskiy) [reply temporal-sdk/src/main/java/io/temporal/internal/sync/SyncWorkflowContext.java:163]: I actually saved the behavior that already was here. It was a null check inside a setter method which is even worse. Probably it was done to avoid double initialization with different implementation, but I'm not sure. But I agree it looks off. Let me try to rework it to throwing an exception if the interceptor is already initialized and see if anything blows up.

File: temporal-sdk/src/test/java/io/temporal/internal/sync/PromiseTest.java (hunk 8 above)
Comment (Spikhalskiy) [reply temporal-sdk/src/test/java/io/temporal/internal/sync/PromiseTest.java:58]: Yeah, but it's actually not what's going on here.
We had some leftover stuff when internal WorkflowInternal had methods that were just calling WorkflowThread methods.
So, we had an "internal" API that didn't do anything else other than calling a "public" API. So, I just removed these methods and replaced them with using an existing public API.

File: temporal-sdk/src/main/java/io/temporal/internal/sync/SyncWorkflowContext.java (hunk 5 above)
Comment (Spikhalskiy) [reply temporal-sdk/src/main/java/io/temporal/internal/sync/SyncWorkflowContext.java:165]: changed to a throwing an exception. see the explanation behind.

File: temporal-sdk/src/main/java/io/temporal/internal/sync/SyncWorkflowContext.java (hunk 4 above)
Comment (Spikhalskiy) [reply temporal-sdk/src/main/java/io/temporal/internal/sync/SyncWorkflowContext.java:157]: changed to a throwing an exception. see the explanation behind.

File: temporal-sdk/src/main/java/io/temporal/internal/sync/BaseRootWorkflowInboundCallsInterceptor.java [hunk 9]
```diff
@@ -0,0 +1,61 @@
...
+  }
+
+  @Override
+  public void init(WorkflowOutboundCallsInterceptor outboundCalls) {
+    workflowContext.initHeadOutboundCallsInterceptor(outboundCalls);
+  }
+
+  @Override
+  public WorkflowOutput execute(WorkflowInput input) {
+    throw new UnsupportedOperationException();
```
Comment (vitarb) [inline temporal-sdk/src/main/java/io/temporal/internal/sync/BaseRootWorkflowInboundCallsInterceptor.java:39]: nit. Pass a human readable message as a parameter. Something like "Root workflow should never be executed directly".

File: temporal-sdk/src/main/java/io/temporal/internal/sync/BaseRootWorkflowInboundCallsInterceptor.java [hunk 10]
```diff
@@ -0,0 +1,61 @@
...
+ *  express or implied. See the License for the specific language governing
+ *  permissions and limitations under the License.
+ */
+
+package io.temporal.internal.sync;
+
+import io.temporal.common.interceptors.WorkflowInboundCallsInterceptor;
+import io.temporal.common.interceptors.WorkflowOutboundCallsInterceptor;
+
+public class BaseRootWorkflowInboundCallsInterceptor implements WorkflowInboundCallsInterceptor {
```
Comment (vitarb) [inline temporal-sdk/src/main/java/io/temporal/internal/sync/BaseRootWorkflowInboundCallsInterceptor.java:25]: I think it would be good to explain in the comment what the root workflow is and why/when this interceptor could be useful.

File: temporal-sdk/src/main/java/io/temporal/internal/sync/WorkflowThreadImpl.java [hunk 11]
```diff
@@ -394,8 +394,15 @@ public void addStackTrace(StringBuilder result) {
...
     int omitTop = 5;
     int omitBottom = 7;
+    // TODO it's not a good idea to rely on the name to understand the thread type. Instead of that
+    // we would better
+    // assign an explicit enum to the threads. This will be especially important when we refactor
+    // root and workflow-method
+    // thread names into names that will include workflowId
     if (DeterministicRunnerImpl.WORKFLOW_ROOT_THREAD_NAME.equals(getName())) {
       omitBottom = 11;
+    } else if (DeterministicRunnerImpl.WORKFLOW_MAIN_THREAD_NAME.equals(getName())) {
```
Comment (vitarb) [inline temporal-sdk/src/main/java/io/temporal/internal/sync/WorkflowThreadImpl.java:404]: nit. You could use `||` instead of `if else if` since then part is the same.
